CRM/Report/Form/Contact/Detail.php

```diff
@@ -975,4 +976,44 @@ public function alterComponentDisplay(&$componentRows) {
     }
   }
 
+  /**
+   * Function to alter from clause by adding checks for permissoned financial type.
+   *
+   * @param string $component
+   *
+   * @return string
+   */
+  public function addPermissionedFromClause($component) {
+    if (is_null($this->financialTypes)) {
+      return NULL;
+    }
+    $financialTypes = 0;
+    if (!empty($this->financialTypes)) {
+      $financialTypes = implode(',', array_keys($this->financialTypes));
+    }
+    $clause = "financial_type_id IN ({$financialTypes})";
+    $whereClause = NULL;
+
+    switch ($component) {
```

Comment:
Passing in component and using it for switch may indicate issue with coupling. On the other hand we need to avoid copy and paste code appearing that will do similar work for each of contribute, membership and event, so we want to use a single function for all three where possible. Not having examined code here in depth, it seems like refactoring to have three functions that marshall arguments that call a single shared function that does not take $component or a proxy for it is one approach. Another is to have a shared function that marshalls arguments that call out to component specific functions. And sometimes we can have 3 component-specific functions that don't have the smell of copy-and-paste. I think we're maybe feeling the effects of a poorly designed class hierarchy here.